### PR TITLE
Vision framework demo 

### PR DESCRIPTION
**Pull request by dorypiacek**
Tutorial used: https://developer.apple.com/documentation/vision/classifying_images_with_vision_and_core_ml

### CODE REVIEW

File: VisionDemo/Podfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+platform :ios, '14.0'
+
+target 'VisionDemo' do
```
Comment (lipovacjetu) [inline VisionDemo/Podfile:3]: Why not Swift Package Manager? 
What's the important difference between the two?

File: .gitignore [hunk 2]
```diff
@@ -54,7 +54,9 @@ playground.xcworkspace
 # you should judge for yourself, the pros and cons are mentioned at:
 # https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control
 #
-# Pods/
+
+Pods/
```
Comment (lipovacjetu) [inline .gitignore:58]: Comment above advises you to include Pods in git (and don't include it in .gitignore).
Why did you decide to go against the commet above :) ?

Comment (lipovacjetu) [reply .gitignore:58]: ➕  for adding .gitignore btw

File: VisionDemo/VisionDemo.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:5]: Hm why is this here?

File: VisionDemo/VisionDemo/Common/Extensions/Array+UniqueElements.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
+//
+//  Array+UniqueElements.swift
+//  VisionDemo
+//
+//  Created by Dory on 26/11/2021.
+//
+
+extension Array where Element: Equatable {
+  func uniqueElements() -> [Element] {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Extensions/Array+UniqueElements.swift:9]: Could we use Set?

File: VisionDemo/VisionDemo/Common/Extensions/CGImagePropertyOrientation+UIImageOrientation.swift [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+//  VisionDemo
+//
+//  Created by Dory on 25/11/2021.
+//
+
+import UIKit
+import ImageIO
+
+/// Source: https://developer.apple.com/documentation/coreml/classifying_images_with_vision_and_core_ml/
+extension CGImagePropertyOrientation {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Extensions/CGImagePropertyOrientation+UIImageOrientation.swift:12]: A-ha! Important consideration when working with UIImage/CIImage... It caused so many headaches in my life past few years :D

File: VisionDemo/VisionDemo/Common/Extensions/UIButton+Closure.swift [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+///
+/// - Parameters:
+///   - owner: Owner to track.
+///   - closure: Closure to execute.
+public func onDealloc(of owner: Any, closure: @escaping () -> Void) {
+	while true {
+		// Generates random key for association and checks that it wasn't used already
+		if let key = UnsafeRawPointer(bitPattern: UInt(arc4random())), objc_getAssociatedObject(owner, key) == nil {
+			let tracker = DeallocTracker(onDealloc: closure)
+			objc_setAssociatedObject(owner, key, tracker, .OBJC_ASSOCIATION_RETAIN)
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Extensions/UIButton+Closure.swift:112]: Do you have experience working with Obj-C runtime and could you explain what exactly happens here? 
How do you feel about using this kind of Obj-C black magic in your Swift code?

File: VisionDemo/VisionDemo/Common/ML/ImageProcessor.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+//  VisionDemo
+//
+//  Created by Dory on 26/11/2021.
+//
+
+import Vision
+import UIKit
+import Combine
+
+final class ImageProcessor {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/ML/ImageProcessor.swift:12]: I like this class!

File: VisionDemo/VisionDemo/Common/ML/ImageProcessor.swift [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		guard let cgImage = image.cgImage else {
+			fatalError("Photo doesn't have underlying CGImage.")
+		}
+
+		let imageClassificationRequest = makeRequest()
+
+		let handler = VNImageRequestHandler(cgImage: cgImage, orientation: orientation)
+		let requests: [VNRequest] = [imageClassificationRequest]
+
+		try handler.perform(requests)
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/ML/ImageProcessor.swift:44]: Are these requests running on the main thread?
What happens if they are and if I pass in a very very very large photo as input?

File: VisionDemo/VisionDemo/Common/Model/ProcessedImage.swift [hunk 9]
```diff
@@ -0,0 +1,27 @@
+//
+//  ProcessedImage.swift
+//  VisionDemo
+//
+//  Created by Dory on 24/11/2021.
+//
+
+import RealmSwift
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Model/ProcessedImage.swift:8]: Why did you choose Realm for storing state? I think your solution is elegant and simple, I'm just wondering what was the reasoning. 

What other options did you have?

File: VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+//
+
+import RealmSwift
+
+class RealmDataProvider {
+	static var shared = RealmDataProvider()
+	
+	private static var config: Realm.Configuration {
+		var config = Realm.Configuration()
+		config.deleteRealmIfMigrationNeeded = true
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift:15]: I'm super scared of this line 😬

File: VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	func read<T: Object>(type: T.Type, with filter: String? = nil) -> Results<T>? {
+		var objects = realm.objects(type)
+		
+		if let filter = filter {
+			objects = objects.filter(filter)
+		}
+		return objects
+	}
+	
+	func update(with closure: @escaping () -> Void) {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift:49]: Is working with Realm in this class thread safe?

File: VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+		do {
+			try realm.write {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift:51]: Could this block the main thread and cause the app to freeze if we're updating the database with a lot of objects or a very large image data object?

File: VisionDemo/VisionDemo/Common/Theme/LocalizationKit.swift [hunk 13]
```diff
@@ -0,0 +1,39 @@
+//
+//  LocalizationKit.swift
+//  VisionDemo
+//
+//  Created by Dory on 25/11/2021.
+//
+
+struct LocalizationKit {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/Theme/LocalizationKit.swift:8]: What is the difference if we would to use `enum` instead of `struct` here?

File: VisionDemo/VisionDemo/Common/UI Components/PrimaryButton.swift [hunk 14]
```diff
@@ -0,0 +1,60 @@
+//
+//  PrimaryButton.swift
+//  VisionDemo
+//
+//  Created by Dory on 25/11/2021.
+//
+
+import UIKit
+
+final class PrimaryButton: UIButton {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/UI Components/PrimaryButton.swift:10]: Ah, did you consider using `https://developer.apple.com/documentation/uikit/uibutton/configuration?changes=_4`?
Demo app runs on iOS15 only, and setting content insets on UIButton is deprecated from that version on, if this UIButton.Configuration is available to you

File: VisionDemo/VisionDemo/Common/UI Components/Tags/TagButton.swift [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	private func setupUI() {
+		backgroundColor = .white
+		
+		setTitleColor(.systemTeal, for: .selected)
+		setTitleColor(.white, for: .normal)
+		
+		layer.borderWidth = Metrics.borderWidth
+		layer.borderColor = UIColor.white.cgColor
+		
+		titleLabel?.font = UIFont.systemFont(ofSize: 13)
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Common/UI Components/Tags/TagButton.swift:57]: should this font size also go to Metrics?

File: VisionDemo/VisionDemo/Modules/Gallery/ViewController/GalleryViewController.swift [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+		if let image = collectionData[indexPath.row].image {
+			cell.update(with: image)
+		}
+		return cell
+	}
+}
+
+// MARK: - UICollectionViewDelegateFlowLayout
+
+extension GalleryViewController: UICollectionViewDelegateFlowLayout {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Modules/Gallery/ViewController/GalleryViewController.swift:157]: This is all good, but a side question -> 
Do you have experience with UICollectionViewCompositionalLayout?

File: VisionDemo/VisionDemo/Modules/Photo/ViewModel/PhotoViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+//
+//  Created by Dory on 25/11/2021.
+//
+
+import Combine
+import Foundation
+import RealmSwift
+
+final class PhotoViewModel {
+	enum Purpose {
```
Comment (lipovacjetu) [inline VisionDemo/VisionDemo/Modules/Photo/ViewModel/PhotoViewModel.swift:13]: Feels like adding this enum and checking what's the Purpose in each method just made things more complicated. Perhaps would be better to have 2 view models?

**Review by lipovacjetu — APPROVED** [review]
Nice! I like the code organization, commit history and some parts of the solution in particular. 
I left some comments on the PR so we have can have a chat about those :)

File: VisionDemo/Podfile (hunk 1 above)
Comment (dorypiacek) [reply VisionDemo/Podfile:3]: CocoaPods has been my dependency manager of choice mostly because of compatibility, but it's true that all frameworks I use are supported by SPM as well. The main difference is that CP includes all pods in a Pods project which is compiled every time I do a clean build.

File: .gitignore (hunk 2 above)
Comment (dorypiacek) [reply .gitignore:58]: 1. Repository size - 600mb including pods vs. 50mb without them 
2. It's pretty simple to install pods with `pod install`
3. Merge conflicts in pods can be a real pain :)

File: VisionDemo/VisionDemo/Common/Extensions/Array+UniqueElements.swift (hunk 4 above)
Comment (dorypiacek) [reply VisionDemo/VisionDemo/Common/Extensions/Array+UniqueElements.swift:9]: Not in my use case, I needed the elements to be in a particular order. Converting to Set doesn't keep the order.

File: VisionDemo/VisionDemo/Common/Model/ProcessedImage.swift (hunk 9 above)
Comment (dorypiacek) [reply VisionDemo/VisionDemo/Common/Model/ProcessedImage.swift:8]: I just think that it's much more simple and easy to use than CoreData :)

File: VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift (hunk 10 above)
Comment (dorypiacek) [reply VisionDemo/VisionDemo/Common/Services/RealmDataProvider.swift:15]: Oh no, I forgot to delete that! 😬 I just used it for testing to prevent the app from crashing when I changed the model.

File: VisionDemo/VisionDemo/Common/UI Components/PrimaryButton.swift (hunk 14 above)
Comment (dorypiacek) [reply VisionDemo/VisionDemo/Common/UI Components/PrimaryButton.swift:10]: Actually the deployment target is 14.0, but thanks for the tip!

File: VisionDemo/VisionDemo/Modules/Gallery/ViewController/GalleryViewController.swift (hunk 16 above)
Comment (dorypiacek) [reply VisionDemo/VisionDemo/Modules/Gallery/ViewController/GalleryViewController.swift:157]: I read about it when it first came out, but haven't used it yet.
